### PR TITLE
Split the generated io.quarkus.runner.Feature#beforeAnalysis() method

### PR DESCRIPTION
**Pull request by mkouba**
- to avoid org.objectweb.asm.MethodTooLargeException

See https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Quarkus.20build.20.20failed

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java [hunk 1]
```diff
@@ -481,10 +539,14 @@ public void write(String s, byte[] bytes) {
 
         MethodDescriptor registerSerializationMethod = null;
 
+        MethodCreator registerForReflection = file
+                .getMethodCreator("registerForReflection", void.class, BeforeAnalysisAccess.class)
+                .setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+
         for (Map.Entry<String, ReflectionInfo> entry : reflectiveClasses.entrySet()) {
             MethodCreator mv = file.getMethodCreator("registerClass" + count++, void.class, Feature.BeforeAnalysisAccess.class);
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:547]: nitpick: 
```suggestion
            MethodCreator mv = file.getMethodCreator("registerClass" + count++, void.class, BeforeAnalysisAccess.class);
```

**Review by zakkak — APPROVED** [review]
LGTM, thanks @mkouba . Let's see what the CI says as well :)

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java [hunk 2]
```diff
@@ -227,55 +232,77 @@ public void write(String s, byte[] bytes) {
...
                         tc.load(runtimeInitializedClassBuildItems.get(i).getClassName()), tc.load(false), cl);
-                tc.writeArrayValue(classes, i, clazz);
+                tc.writeArrayValue(classesArray, i, clazz);
                 CatchBlockCreator cc = tc.addCatch(Throwable.class);
                 cc.invokeVirtualMethod(ofMethod(Throwable.class, "printStackTrace", void.class), cc.getCaughtException());
             }
+            runtimeInitializedClasses.returnValue(classesArray);
+
+            ResultHandle classes = overallCatch.invokeStaticMethod(runtimeInitializedClasses.getMethodDescriptor());
             overallCatch.invokeStaticMethod(INITIALIZE_CLASSES_AT_RUN_TIME, classes);
```
[the remark this answers is not in the document]
Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:268]: Which calls do you have in mind? The idea is not to leave out any logic but create a private static method for each logical part of the `beforeAnalysis()` method.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java (hunk 1 above)
Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:547]: We use `Feature.BeforeAnalysisAccess` all around so I wanted to be consistent...

**Comment by mkouba** [comment]
> LGTM, thanks @mkouba . Let's see what the CI says as well :)

Thanks for the review and yes, we'll see if the CI is green :-)

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:268]: You could invoke `INITIALIZE_CLASSES_AT_RUN_TIME` directly in `runtimeInitializedClasses` for instance.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java [hunk 3]
```diff
@@ -481,10 +539,14 @@ public void write(String s, byte[] bytes) {
 
         MethodDescriptor registerSerializationMethod = null;
 
+        MethodCreator registerForReflection = file
+                .getMethodCreator("registerForReflection", void.class, BeforeAnalysisAccess.class)
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:543]:
```suggestion
                .getMethodCreator("registerForReflection", void.class, Feature.BeforeAnalysisAccess.class)
```

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java [hunk 4]
```diff
@@ -17,6 +17,7 @@
 import org.graalvm.home.Version;
 import org.graalvm.nativeimage.ImageSingletons;
 import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.Feature.BeforeAnalysisAccess;
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:20]:
```suggestion
```

**Review by zakkak — CHANGES_REQUESTED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java (hunk 1 above)
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:547]: You are right. So it's the other way around, please see my new suggestions.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:268]: I see. Well, no reason really. It's functionally equivalent and I don't think that it really matters ;-).

**Review by gsmet — APPROVED** [review]
Thanks for taking care of this.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java [hunk 5]
```diff
@@ -214,11 +214,15 @@ public void write(String s, byte[] bytes) {
...
         MethodCreator beforeAn = file.getMethodCreator("beforeAnalysis", "V", BEFORE_ANALYSIS_ACCESS);
         TryBlock overallCatch = beforeAn.tryBlock();
-        //TODO: at some point we are going to need to break this up, as if it get too big it will hit the method size limit
 
         ResultHandle beforeAnalysisParam = beforeAn.getMethodParam(0);
+
+        MethodCreator registerAsUnsafeAccessed = file
+                .getMethodCreator("registerAsUnsafeAccessed", void.class, Feature.BeforeAnalysisAccess.class)
+                .setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+        overallCatch.invokeStaticMethod(registerAsUnsafeAccessed.getMethodDescriptor(), beforeAnalysisParam);
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java:223]: It's a very minor comment but any chance you could follow the same pattern for this one and call the method after the `returnVoid`?
At first I thought something was missing.

It's really minor so feel free to ignore if you think it's not worth it.

**Review by zakkak — APPROVED** [review]
Thanks @mkouba

File: core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java (hunk 5 above)
Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java]: Done.

**Pull request merged**